paymentsheet/src/main/java/com/stripe/android/customersheet/data/CustomerSessionElementsSessionManager.kt

```diff
@@ -0,0 +1,98 @@
+package com.stripe.android.customersheet.data
+
+import com.stripe.android.core.injection.IOContext
+import com.stripe.android.customersheet.CustomerSheet
+import com.stripe.android.customersheet.ExperimentalCustomerSheetApi
+import com.stripe.android.model.ElementsSession
+import com.stripe.android.paymentsheet.ExperimentalCustomerSessionApi
+import com.stripe.android.paymentsheet.PaymentSheet
+import com.stripe.android.paymentsheet.PrefsRepository
+import com.stripe.android.paymentsheet.model.SavedSelection
+import com.stripe.android.paymentsheet.repositories.ElementsSessionRepository
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.coroutines.CoroutineContext
+
+internal interface CustomerSessionElementsSessionManager {
+    suspend fun fetchCustomerSessionEphemeralKey(): Result<CachedCustomerEphemeralKey>
+
+    suspend fun fetchElementsSession(): Result<ElementsSession>
+}
+
+@OptIn(ExperimentalCustomerSheetApi::class, ExperimentalCustomerSessionApi::class)
+@Singleton
+internal class DefaultCustomerSessionElementsSessionManager @Inject constructor(
+    private val elementsSessionRepository: ElementsSessionRepository,
+    private val prefsRepositoryFactory: @JvmSuppressWildcards (String) -> PrefsRepository,
+    private val customerSessionProvider: CustomerSheet.CustomerSessionProvider,
+    private val timeProvider: () -> Long,
+    @IOContext private val workContext: CoroutineContext,
+) : CustomerSessionElementsSessionManager {
+    @Volatile
+    private var cachedCustomerEphemeralKey: CachedCustomerEphemeralKey? = null
+
+    private var intentConfiguration: CustomerSheet.IntentConfiguration? = null
+
+    override suspend fun fetchCustomerSessionEphemeralKey(): Result<CachedCustomerEphemeralKey> {
+        return withContext(workContext) {
+            cachedCustomerEphemeralKey.takeUnless { cachedCustomerEphemeralKey ->
+                cachedCustomerEphemeralKey == null || cachedCustomerEphemeralKey.shouldRefresh(
+                    timeProvider()
+                )
+            }?.let {
+                Result.success(it)
+            } ?: run {
+                fetchElementsSession().mapCatching {
+                    cachedCustomerEphemeralKey
+                        ?: throw IllegalStateException("Should have been initialized from `elements/session`!")
```

Comment:
I think we should have a better way to represent whether the customer ephemeral key has been loaded than using null. This error could get actually get thrown in two cases:
- some unexpected case where `fetchElementsSession()` doesn't set `cachedCustomerEphemeralKey` correctly
- the elements session `customer` or `customer.session` is null. 

We can't currently tell the difference between these two cases. I think it's important to differentiate here, because our fix would be very different for the the two cases